### ISSUE TITLE
Rust Analyzer detects a syntax error when the code compiles and runs just fine

### ISSUE THREAD

**Issue by khalid-codes** [rust-analyzer/rust-analyzer#10784]
Here is the sample code:
```rust
fn collatz(n: u64) -> Option<u64> {
    const OVERFLOW_BARRIER: u64 = u64::MAX / 3;
    let mut n = n;
    for i in 0.. {
        n = match (n, n % 2) {
            (0, _) => break,
            (1, _) => return Some(i),
            (OVERFLOW_BARRIER.., 1) => 0, // <- Errors out here.
            (_, 0) => n / 2,
            _ => n * 3 + 1,
        }
    }
    None
}
```

Changing `OVERFLOW_BARRIER..` to `OVERFLOW_BARRIER..=u64::MAX` seems to work.

But changing placement of `match` tuple works just fine: 

```rust
pub fn collatz(n: u64) -> Option<u64> {
    const OVERFLOW_BARRIER: u64 = u64::MAX / 3;
    let mut n = n;
    for i in 0.. {
        n = match (n % 2, n) {
            (_, 0) => break,
            (_, 1) => return Some(i),
            (1, OVERFLOW_BARRIER..) => 0, // Detects no error.
            (0, _) => n / 2,
            _ => n * 3 + 1,
        }
    }
    None
}
```

But Rust compiler doesn't detect any error. It compiles just fine.

**Comment by Veykril** [comment]
I think this was reported before but I can't find where.

**Comment by jonas-schievink** [comment]
Maybe in https://github.com/rust-analyzer/rust-analyzer/issues/9066?
